### PR TITLE
Fixed bug of snapshot from chaos test.

### PR DESCRIPTION
**Pull request by bright-starry-sky**
1,  Fixed checkpoint bug. 
2, Refused to create snapshot when index rebuilding.

### CODE REVIEW

**Comment by dangleptr** [comment]
Could you explain the bug in detail ?

File: src/meta/processors/admin/SnapShot.cpp [hunk 1]
```diff
@@ -21,38 +21,42 @@ cpp2::ErrorCode Snapshot::createSnapshot(const std::string& name) {
                    << static_cast<int32_t>(ret);
         return cpp2::ErrorCode::E_STORE_FAILURE;
     }
-    for (auto& space : spaces) {
-        auto status = client_->createSnapshot(space, name).get();
-        if (!status.ok()) {
-            return cpp2::ErrorCode::E_RPC_FAILURE;
+    auto hosts = ActiveHostsMan::getActiveHosts(kv_);
```
Comment (dangleptr) [inline src/meta/processors/admin/SnapShot.cpp:24]: What's the difference with the original logic.

File: src/meta/processors/admin/SnapShot.cpp [hunk 2]
```diff
@@ -77,78 +81,37 @@ bool Snapshot::getAllSpaces(std::vector<GraphSpaceID>& spaces, kvstore::ResultCo
 }
 
 cpp2::ErrorCode Snapshot::blockingWrites(storage::cpp2::EngineSignType sign) {
+    folly::SharedMutex::ReadHolder rHolder(LockUtils::spaceLock());
+    // check the index rebuild. not allowed to block write when index rebuilding.
+    auto prefix = MetaServiceUtils::rebuildIndexStatusPrefix();
```
Comment (dangleptr) [inline src/meta/processors/admin/SnapShot.cpp:86]: Why not put the check inside processor.

File: src/meta/processors/admin/SnapShot.cpp [hunk 3]
```diff
@@ -77,78 +81,37 @@ bool Snapshot::getAllSpaces(std::vector<GraphSpaceID>& spaces, kvstore::ResultCo
...
-
-        if (!status.ok()) {
-            return cpp2::ErrorCode::E_BLOCK_WRITE_FAILURE;
-        }
-    }
-    return cpp2::ErrorCode::SUCCEEDED;
-}
-
-std::unordered_map<HostAddr, std::vector<PartitionID>>
-Snapshot::getLeaderParts(HostLeaderMap *hostLeaderMap, GraphSpaceID spaceId) {
```
Comment (dangleptr) [inline src/meta/processors/admin/SnapShot.cpp:98]: We don't need the method any more?

File: src/meta/processors/admin/AdminClient.cpp [hunk 4]
```diff
@@ -603,15 +604,15 @@ folly::Future<Status> AdminClient::createSnapshot(GraphSpaceID spaceId, const st
      * Don't need retry.
      * Because existing checkpoint directories leads to fail again.
      **/
-    getResponse(std::move(allHosts), 0, std::move(req), [] (auto client, auto request) {
+    getResponse({host}, 0, std::move(req), [] (auto client, auto request) {
```
Comment (darionyaphet) [inline src/meta/processors/admin/AdminClient.cpp:607]: Seems you will create snapshot at host one by one?

**Comment by bright-starry-sky** [comment]
> Could you explain the bug in detail ?
A very humiliating mistake, the error leads to lost checkpoint in some storage node. 
I misunderstood the meaning of 'getResponse'

File: src/meta/processors/admin/AdminClient.cpp (hunk 4 above)
Comment (bright-starry-sky) [reply src/meta/processors/admin/AdminClient.cpp:607]: Yes, you are right.

File: src/meta/processors/admin/SnapShot.cpp (hunk 3 above)
Comment (bright-starry-sky) [reply src/meta/processors/admin/SnapShot.cpp:98]: Yes,This method is no longer used.

File: src/meta/processors/admin/SnapShot.cpp (hunk 2 above)
Comment (bright-starry-sky) [reply src/meta/processors/admin/SnapShot.cpp:86]: Good point.

File: src/meta/processors/admin/SnapShot.cpp [hunk 5]
```diff
@@ -77,78 +81,37 @@ bool Snapshot::getAllSpaces(std::vector<GraphSpaceID>& spaces, kvstore::ResultCo
...
 cpp2::ErrorCode Snapshot::blockingWrites(storage::cpp2::EngineSignType sign) {
+    folly::SharedMutex::ReadHolder rHolder(LockUtils::spaceLock());
+    // check the index rebuild. not allowed to block write when index rebuilding.
+    auto prefix = MetaServiceUtils::rebuildIndexStatusPrefix();
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto ret = kv_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+    if (ret != kvstore::ResultCode::SUCCEEDED) {
+        return cpp2::ErrorCode::E_BLOCK_WRITE_FAILURE;
+    }
+    while (iter->valid()) {
```
Comment (dangleptr) [inline src/meta/processors/admin/SnapShot.cpp:92]: why use while here?

Comment (bright-starry-sky) [reply src/meta/processors/admin/SnapShot.cpp:92]: > why use while here?

I need check whether any index is being rebuild for all spaces and all indexes.

**Comment by bright-starry-sky** [comment]
Address's comments .move the index check logic to processor.

**Pull request merged**